Add docs for some methods in vim.lsp

runiq opened the pull request:
Some remaining issues:

- The functions in `vim.lsp.callbacks` don't show up because of their, uh, idiosyncratic naming scheme. I had to make some tweaks to `scripts/gen_vimdoc.py` to keep it running in the absence of any documented method in that module.
- I made the mistake of starting with `lsp.lua`, which contains the LSP client object, which is arguably the most important bit of the whole thing. I'm not sure if I've done everything correctly in there. I went over it again after going through the rest of the modules, but I'm pretty sure I messed something up or missed something. It's time for someone else's eyeballs on this.
- Is there a way to expose non-function values in the docs, for example the big table in `vim.lsp.protocol`?

teto commented:
cc @tjdevries

tjdevries commented:
I think `@private` will make them not show up in the docs. Will review later.

KillTheMule commented:
> I think `@private` will make them not show up in the docs

True, we're using doxygen, which has the [private](https://www.doxygen.nl/manual/commands.html#cmdprivate) magic command. It's converted [here](https://github.com/neovim/neovim/blob/master/scripts/lua2dox.lua#L410), i.e. you'll need to use `--@private`.

runiq commented:
Cool, I'll change those, thanks!

runiq commented:
Alright, I think I'm done for now. Everything public in `vim.lsp` appears to be documented now, and some private stuff is too.

cc @tjdevries

runiq commented:
Not sure what's up with the Appveyor test here…

tjdevries commented on runtime/lua/vim/lsp.lua:
```diff
@@ -488,14 +586,39 @@ function lsp.start_client(config)
   end
 
-  --- Checks capabilities before rpc.request-ing.
+  --@private
```
I'd have to check again whether this is actually the case, but can't users access all of the `client.*` style functions from the returned client object?

runiq replied:
Yeah, AFAIK they can. (Are you asking why I marked all the `client.` methods as private? These methods are exposed in the vim docs with the `client.` prefix stripped off, and  I couldn't get lua2dox/gen_vimdoc.py to change that. So I just marked them as private and decided to document them on the client() function instead.)

runiq replied:
The reason why I documented the `client.` functions in two locations is that Lua langservers take their docs from the function docstrings, while the vimdocs are generated from the `client()` object, so I'll have to keep both in sync (for now).

tjdevries replied:
Hmmm OK. Maybe just make a note of that (for future readers)

runiq replied:
Will do.

Is there anything else I should change about the functions on the client object? That's definitely the part I was the most unsure about.

runiq replied:
Added a note and cleaned up the docs on the client object somewhat.

tjdevries commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -104,6 +141,13 @@ function M.formatting_sync(options, timeout_ms)
+--- Perform |vim.lsp.buf.formatting()| synchronously.
+---
+--- Useful for running on save, to make sure buffer is formatted prior to being
+--- saved. {timeout_ms} is passed on to |vim.lsp.buf_request_sync()|.
```
Maybe we could just add an example with the `<pre>` style w/ an autocommand to show how to do it?

runiq replied:
Sure, will do.

tjdevries commented on runtime/lua/vim/lsp/log.lua:
```diff
@@ -15,18 +15,53 @@ log.levels = {
+--- Recommended usage:
+---
+---<pre>
+---    local _ = log.debug() and log.debug("123")
```
Can we change this? We should have two examples:

The first is with a static string (like above) then you can do `log.debug("123")`. No need to do the `and` check.

If it's an `expensive_function`, then you can do this strategy (so that Lua does not execute expensive function.

In general, it is not necessary to do the `and` check.

runiq replied:
Will do. I also forgot documenting params and return values…

tjdevries commented on runtime/lua/vim/lsp/log.lua:
```diff
@@ -74,6 +99,8 @@ end
 vim.tbl_add_reverse_lookup(log.levels)
 
+--- Sets the current log level.
+--@param level (string or number) One of `vim.lsp.log.levels`
```
Should probably be a tag (do we have a tag for the levels?)

runiq replied:
Not yet, I can add one. In this PR?

tjdevries replied:
Yeah. You can hard code some things into the docs if you put them above the generated code.  Or maybe at the top w/ a `@brief`? I would have to experiment with that to check though.

runiq replied:
Hmm, I'd rather not add stuff manually to the doc output, it could easily disappear if someone doesn't look too hard at a commit's changes.

Is there maybe already some easy way to get "module documentation" to show in the Vimdoc output? (cc @KillTheMule)

runiq replied:
cc @justinmk:

> Is there maybe already some easy way to get "module documentation" to show in the Vimdoc output?

tjdevries commented on runtime/lua/vim/lsp/rpc.lua:
```diff
@@ -294,7 +392,14 @@ local function create_and_start_client(cmd, cmd_args, handlers, extra_spawn_para
   end
 
-  local function send_request(method, params, callback)
+  -- FIXME: Should be placed on the RPC client object returned by `start()`
```
Yeah, some of this I hope to one day handle with tree sitter generated docs...

runiq replied:
I am anxiously awaiting this, yeah :XD:

tjdevries commented on runtime/doc/lsp.txt:
```diff
@@ -739,28 +674,38 @@ resolve_capabilities({server_capabilities})
 
 code_action({context})                             *vim.lsp.buf.code_action()*
-                TODO: Documentation
+                Select a code action from the input list tha tis available at
```
typo "tha tis"

tjdevries commented:
Overall, it looks really good. I'm more of a mind that we can merge this sooner (rather than waiting forever) and try and do more incremental updates of the docs as we go forward.... `TODO: Documentation` on a bunch of functions that don't exist is not very helpful for anyone :laughing:

runiq committed:
gen_vimdoc: Allow to keep intermediary output

runiq commented:
Alright, addressed all the review comments I could. Unfortunately I didn't find a way to massage doxygen/lua2dox/gen_vimdoc.py output so that I could output the correct function names for the client object. I also didn't manage to add help tags for the log levels, because everything I tried looked fairly ugly. :-/

runiq committed:
Add docs for most vim.lsp methods

Most of the lsp.log will be addressed in a separate PR.

runiq commented:
@tjdevries Oh, another issue I just realized: I'm not sure I completely understand the handlers vs. callbacks distinction on the client object. Could you maybe check those a little closer?

Thanks!

tjdevries reviewed (approved):
As mentioned, there are some FIXMEs for later, but great progress.

runiq committed:
Add FIXMEs

tjdevries commented:
We can update the docs if anyone who has refrained from commenting has complaints :) This is a very strict & large improvement over the current situation.

justinmk commented:
> I think `@private` will make them not show up in the docs. Will review later.

also use a `_` prefix on function names, which imo is preferred because it gives a hint when reading code. (and could be useful for syntax highlighting in theory)

tjdevries commented:
Yes, agreed :) But I wanted to make the docs PR as non-invasive as possible since I knew the diff would be huge :laughing: we can fix up the functions later (or, we can get smarter generators.... :wink: hopefully some time).
